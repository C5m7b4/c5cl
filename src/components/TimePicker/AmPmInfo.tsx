import React from 'react';
import { pad } from '../../utils';

export interface AmPMInfoProps {
  time: Date;
  onClick: (b: boolean) => void;
}

const AmPmInfo = (props: AmPMInfoProps) => {
  let am = true;
  if (props.time.getHours() >= 12) {
    am = false;
  }
  const getTimeString = (hour: number, minute: number) => {
    let am = true;
    if (hour >= 12) {
      am = false;
    }
    return (
      hour + ':' + pad(minute.toString(), 2, '0') + ' ' + (am ? 'AM' : 'PM')
    );
  };
  const { time } = props;
  return (
    <div className="am-pm-info">
      <div
        className={'am' + (am ? ' selected' : '')}
        onClick={() => props.onClick(true)}
      >
        AM
      </div>
      <div className="time">
        {getTimeString(time.getHours(), time.getMinutes())}
      </div>
      <div
        className={'pm' + (!am ? ' selected' : '')}
        onClick={() => props.onClick(false)}
      >
        PM
      </div>
    </div>
  );
};

export default AmPmInfo;
