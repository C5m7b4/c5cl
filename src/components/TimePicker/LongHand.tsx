import React from 'react';

export interface LongHandProps {
  selected: number;
  type: string;
}

const LongHand = (props: LongHandProps) => {
  const deg = (props.selected / (props.type === 'hours' ? 12 : 60)) * 360;
  return (
    <div>
      <div
        className="long-hand"
        style={{
          transform: `rotate(${deg}deg) scale(1, 0.8)`,
          WebkitTransform: `rotate(${deg}deg) scale(1, 0.8)`,
        }}
      ></div>
      <div className="long-hand-attachment"></div>
    </div>
  );
};

export default LongHand;
