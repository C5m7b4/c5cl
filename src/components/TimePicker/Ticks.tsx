import React from 'react';

const Ticks = () => {
  const buildTick = (index: number) => {
    return React.createElement(
      'div',
      {
        key: index,
        className: 'tick ' + (index % 5 === 0 ? 'big ' : ''),
        style: {
          transform: 'rotate(' + index * 6 + 'deg)',
          WebkitTransform: 'rotate(' + index * 6 + 'deg)',
        },
      },
      React.createElement('div', null)
    );
  };

  const buildTicks = () => {
    const ticks: any[] = [];
    for (var i = 0; i < 60; i++) {
      ticks.push(buildTick(i));
    }
    return ticks;
  };

  return <div className="ticks">{buildTicks()}</div>;
};

export default Ticks;
