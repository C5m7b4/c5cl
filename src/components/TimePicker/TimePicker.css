.c5-time-picker {
  position: relative;
}

.c5-time-picker * {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5-tp-input {
  width: 150px;
  height: 35px;
  overflow: hidden;
  border-radius: 20px;
}

.c5-tp-input:focus {
  outline: none;
}

.c5-tp-input input {
  width: 125%;
  background: #f5f5f5;
  border: 0;
  outline: none;
  height: 35px;
  text-align: center;
  cursor: pointer;
  user-select: none;
}

.mtp-container {
  position: absolute;
  left: 0;
  top: 40px;
  width: 160px;
  min-height: 150px;
  background: #fff;
  box-shadow: 10px 10px 40px rgba(0, 0, 0, 0.3);
  border-radius: 20px;
  overflow: hidden;
  padding: 25px 30px;
}

.clock-face-wrapper {
  position: relative;
  background: white;
  border: solid 1px #dadada;
  border-radius: 100%;
  height: 151px;
  width: 151px;
}

/* clock positions */
div.c5-time-picker div.clock div.face.face-show {
  visibility: visible;
  opacity: 1;
  transition: opacity 350ms 175ms ease;
}

div.c5-time-picker div.clock div.face {
  position: absolute;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 3;
  background: white;
}

div.c5-time-picker div.clock div.face div.position {
  position: absolute;
  color: black;
  font-size: 0.8em;
  z-index: 7;
  border-radius: 100%;
  cursor: pointer;
  width: 29px;
  display: inline-block;
  height: 29px;
  line-height: 29px;
  vertical-align: middle;
  text-align: center;
}

div.c5-time-picker div.clock div.face div.position.position-1 {
  left: 87px;
  top: 12px;
}

div.c5-time-picker div.clock div.face div.position.position-2 {
  left: 110px;
  top: 32px;
}

div.c5-time-picker div.clock div.face div.position.position-3 {
  left: 118px;
  top: 60.5px;
}

div.c5-time-picker div.clock div.face div.position.position-4 {
  left: 110px;
  top: 89px;
}

div.c5-time-picker div.clock div.face div.position.position-5 {
  left: 87px;
  top: 111px;
}

div.c5-time-picker div.clock div.face div.position.position-6 {
  left: 60.5px;
  top: 117px;
}

div.c5-time-picker div.clock div.face div.position.position-7 {
  left: 34px;
  top: 111px;
}

div.c5-time-picker div.clock div.face div.position.position-8 {
  left: 12px;
  top: 89px;
}

div.c5-time-picker div.clock div.face div.position.position-9 {
  left: 4px;
  top: 60.5px;
}
div.c5-time-picker div.clock div.face div.position.position-10 {
  left: 12px;
  top: 32px;
}

div.c5-time-picker div.clock div.face div.position.position-11 {
  left: 34px;
  top: 12px;
}

div.c5-time-picker div.clock div.face div.position.position-12 {
  left: calc(50% - 14.5px);
  top: 4px;
}

/* ampm info */
div.c5-time-picker div.clock div.am-pm-info {
  float: left;
  width: 100%;
  margin-top: 16px;
}

div.c5-time-picker div.clock div.am-pm-info div.am.selected {
  background: #026aa7;
  color: white;
}

div.c5-time-picker div.clock div.face.face-hide {
  visibility: hidden;
  opacity: 0;
  transition: visibility 0s 350ms, opacity 350ms ease;
}

div.c5-time-picker div.clock div.face div.inner-face {
  position: absolute;
  z-index: 5;
  width: 79px;
  height: 79px;
  border: solid 1px #dadada;
  top: 35px;
  left: 35px;
  border-radius: 100%;
  background: white;
}

div.c5-time-picker div.clock div.face div.position.selected {
  background: #026aa7;
  color: white;
}

div.c5-time-picker div.clock.clock-hide .position,
div.c5-time-picker div.clock.clock-hide .inner-face,
div.c5-time-picker div.clock.clock-hide .long-hand,
div.c5-time-picker div.clock.clock-hide .long-hand-attachment,
div.c5-time-picker div.clock.clock-hide .ticks {
  transition: visibility 0s 350ms;
  visibility: hidden;
}

div.c5-time-picker div.clock div.face div.long-hand {
  height: 75px;
  width: 1px;
  background: #026aa7;
  position: absolute;
  left: 50%;
  top: 0;
  z-index: 6;
  -webkit-transform-origin: 50% 100%;
  transform-origin: 50% 100%;
}

div.c5-time-picker div.clock div.face div.long-hand-attachment {
  width: 7px;
  height: 7px;
  border-radius: 100%;
  z-index: 7;
  background: #026aa7;
  position: absolute;
  top: 72px;
  left: 73px;
}

div.c5-time-picker div.clock div.face div.ticks div.tick {
  height: 75px;
  width: 1px;
  background: white;
  position: absolute;
  left: 50%;
  top: 0;
  z-index: 3;
  -webkit-transform-origin: 0% 100%;
  transform-origin: 0% 100%;
}
div.c5-time-picker div.clock div.face div.ticks div.tick div {
  position: absolute;
  z-index: 4;
  background: #aaa;
  top: 0;
  height: 4px;
  width: 1px;
}
div.c5-time-picker div.clock div.face div.ticks div.tick.big div {
  height: 7px;
}

/* am-pm-info */
div.c5-time-picker div.clock div.am-pm-info {
  float: left;
  width: 100%;
  margin-top: 16px;
}
div.c5-time-picker div.clock div.am-pm-info div.am,
div.c5-time-picker div.clock div.am-pm-info div.pm {
  float: left;
  color: black;
  border: solid 1px #dadada;
  font-size: 0.7em;
  padding: 4px 7px;
  border-radius: 3px;
  background: white;
  cursor: pointer;
  transition: all .3s ease;
}
div.c5-time-picker div.clock div.am-pm-info div.am:hover,
div.c5-time-picker div.clock div.am-pm-info div.pm:hover {
  background: #eaeaea;
}
div.c5-time-picker div.clock div.am-pm-info div.am.selected,
div.c5-time-picker div.clock div.am-pm-info div.pm.selected {
  background: #026aa7;
  color: white;
}
div.c5-time-picker div.clock div.am-pm-info div.pm {
  float: right;
}
div.c5-time-picker div.clock div.am-pm-info div.time {
  float: left;
  color: #777;
  font-size: 0.9em;
  width: calc(100% - 68px);
  text-align: center;
  display: inline-block;
  vertical-align: middle;
  height: 23px;
  line-height: 23px;
}

div.c5-time-picker div.clock div.close {
  float: left;
  display: flex;
  justify-content: center;
  margin-top: 10px;
  align-items: center;
  width: 100%;
}

div.c5-time-picker div.clock div.close button {
  width: 100%;
  background-color: #026aa7;
  color: white;
  border-radius: 10px;
  border: 0;
  outline: none;
  padding: 8px 10px;
  cursor: pointer;
  transition: all .3s ease;
}

div.c5-time-picker div.clock div.close button:hover{
  background-color: #3589ba;
}

div.c5-time-picker div.clock div.close button:active{
  transform: scale(0.96);
}
