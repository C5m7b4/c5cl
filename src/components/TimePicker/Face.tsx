import React, { useRef } from 'react';
import LongHand from './LongHand';
import Ticks from './Ticks';

interface FaceProps {
  type: string;
  visible: boolean;
  values: string[];
  selected: number;
  onClick: (h: string) => void;
}

const Face = (props: FaceProps) => {
  const faceRef = useRef<HTMLDivElement>(null);
  return (
    <div
      ref={faceRef}
      className={`face ${props.type} ${
        props.visible ? ' face-show' : 'face-hide'
      }`}
    >
      {props.values.map((v, i) => {
        return (
          <div
            key={`face-${i}`}
            className={
              'position position-' +
              (i + 1) +
              (parseInt(props.selected.toString()) === parseInt(v)
                ? ' selected'
                : '')
            }
            onClick={() => props.onClick(v)}
          >
            {v}
          </div>
        );
      })}
      <LongHand type={props.type} selected={props.selected} />
      <div className="inner-face"></div>
      <Ticks />
    </div>
  );
};

export default Face;
