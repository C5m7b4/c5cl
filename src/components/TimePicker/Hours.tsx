import React from 'react';
import Face from './Face';

const Hours = (props: any) => {
  const { time } = props;

  const buildHours = () => {
    var hours = [];
    for (var i = 1; i <= 12; i++) {
      hours.push(i);
    }
    return hours;
  };

  return (
    <Face
      {...props}
      type="hours"
      values={buildHours()}
      selected={time.getHours() >= 12 ? time.getHours() - 12 : time.getHours()}
    />
  );
};

export default Hours;
