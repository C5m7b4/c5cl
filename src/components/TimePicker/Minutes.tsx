import React from 'react';
import Face from './Face';
import { pad } from '../../utils';

const Minutes = (props: any) => {
  const { time } = props;

  const buildMinutes = () => {
    var minutes = [];
    for (var i = 1; i <= 12; i++) {
      let val = (i === 12 ? 0 : i) * 5;
      minutes.push(pad(val.toString(), 2, '0'));
    }
    return minutes;
  };

  return (
    <Face
      {...props}
      type="minutes"
      values={buildMinutes()}
      selected={props.time.getMinutes()}
    />
  );
};

export default Minutes;
