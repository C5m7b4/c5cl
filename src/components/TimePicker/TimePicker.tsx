import { hasSelectionSupport } from '@testing-library/user-event/dist/utils';
import React, { useState, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import Hours from './Hours';
import Minutes from './Minutes';
import AmPmInfo from './AmPmInfo';
import { pad } from '../../utils';

import './TimePicker.css';

export type TimePickerReturnValue = Date | string;

export interface TimePickerProps {
  time: Date;
  onChange: (t: TimePickerReturnValue) => void;
  id?: string;
  style?: React.CSSProperties;
  returnShortTime?: boolean;
}

const TimePicker = (props: TimePickerProps) => {
  const [defaultTime, setDefaultTime] = useState(props.time);
  const [showTimePicker, setShowTimePicker] = useState(false);
  const [showHours, setShowHours] = useState(true);
  const [showMinutes, setShowMinutes] = useState(false);

  const timeRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    setInput();
    return () => {
      window.removeEventListener('click', addBackdrop, true);
    };
  }, [defaultTime]);

  const formatTime = () => {
    const newDate = new Date(defaultTime);
    let hours = newDate.getHours();
    const minutes = newDate.getMinutes();
    return (
      pad(hours.toString(), 2, '0') + ':' + pad(minutes.toString(), 2, '0')
    );
  };

  const setInput = () => {
    const newDate = formatTime();
    if (timeRef.current) {
      timeRef.current.value = newDate;
    }
  };

  const componentId = props.id
    ? `c5-time-picker-${props.id}`
    : 'c5-time-picker-1';

  const handleTimeClick = () => {
    setShowTimePicker(true);
    window.addEventListener('click', addBackdrop, true);
  };

  const addBackdrop = (e: MouseEvent) => {
    const div = document.querySelector('.mtp-container');
    if (!div) {
      setShowTimePicker(false);
      window.removeEventListener('click', addBackdrop, true);
      return;
    }
    const rect = div.getBoundingClientRect();
    if (e.clientX < rect.left || e.clientX > rect.right) {
      setShowTimePicker(false);
      window.removeEventListener('click', addBackdrop, true);
      return;
    }
    if (e.clientY < rect.top || e.clientY > rect.bottom) {
      setShowTimePicker(false);
      window.removeEventListener('click', addBackdrop, true);
      return;
    }
  };

  const getTimePieces = () => {
    let hour = defaultTime.getHours();
    let minute = defaultTime.getMinutes();
    let ampm = 'AM';
    if (hour >= 12) {
      hour = hour - 12;
      ampm = 'PM';
    }
    let month = defaultTime.getMonth() + 1;
    let day = defaultTime.getDate();
    let year = defaultTime.getFullYear();

    return {
      hour,
      minute,
      month,
      day,
      year,
      ampm,
    };
  };

  const handleHoursChange = (h: number) => {
    const { minute, month, day, year, ampm } = getTimePieces();
    const newDate = new Date(
      month + '/' + day + '/' + year + ' ' + h + ':' + minute + ' ' + ampm
    );
    setShowHours(false);
    setShowMinutes(true);
    setDefaultTime(newDate);
  };

  const handleMinutesChange = (m: number) => {
    const { hour, month, day, year, ampm } = getTimePieces();
    const newDate = new Date(
      month + '/' + day + '/' + year + ' ' + hour + ':' + m + ' ' + ampm
    );
    setDefaultTime(newDate);
  };

  const handleClose = () => {
    if (props.onChange && !props.returnShortTime) {
      props.onChange(defaultTime);
    }
    if (props.onChange && props.returnShortTime) {
      const hour = defaultTime.getHours();
      const minute = defaultTime.getMinutes();
      let ampm = 'AM';
      if (hour >= 12) {
        ampm = 'PM';
      }
      props.onChange(hour + ':' + pad(minute.toString(), 2, '0') + ' ' + ampm);
    }
    setShowTimePicker(false);
  };

  const handleAmPmChange = (b: boolean) => {
    let ampm = 'AM';
    if (!b) {
      ampm = 'PM';
    }
    const { hour, minute, month, day, year } = getTimePieces();
    const newDate = new Date(
      month + '/' + day + '/' + year + ' ' + hour + ':' + minute + ' ' + ampm
    );
    setDefaultTime(newDate);
  };

  return (
    <div id="c5-time-picker" className="c5-time-picker" style={props.style}>
      <div className="c5-tp-input" onClick={handleTimeClick} id={componentId}>
        <input type="time" ref={timeRef} />
      </div>
      {showTimePicker ? (
        <div className="mtp-container clock">
          <div className="clock-face-wrapper">
            <Hours
              visible={showHours}
              time={defaultTime}
              onClick={handleHoursChange}
              onHidden={showHours}
            />
            <Minutes
              visible={showMinutes}
              time={defaultTime}
              onClick={handleMinutesChange}
              onHidden={showMinutes}
            />
          </div>
          <AmPmInfo time={defaultTime} onClick={handleAmPmChange} />
          <div className="close">
            <button onClick={handleClose}>Close</button>
          </div>
        </div>
      ) : null}
    </div>
  );
};

export default TimePicker;
