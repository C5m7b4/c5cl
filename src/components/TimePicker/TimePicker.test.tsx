import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import TimePicker from './TimePicker';

const dte = new Date('3/20/2022 3:00 AM');

describe('TimePicker', () => {
  test('should render correctly', () => {
    const testFn = jest.fn();
    const { container } = render(<TimePicker time={dte} onChange={testFn} />);
    expect(container).toMatchSnapshot();
  });
  test('should handle active component', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={dte} onChange={testFn} />
    );
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);
  });
  test('should handle pm', () => {
    const testFn = jest.fn();
    const { container } = render(
      <TimePicker time={new Date('3/20/2022 4:00 PM')} onChange={testFn} />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle am click', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={new Date('3/20/2022 4:00 PM')} onChange={testFn} />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const btn1 = container.querySelector(
      '.am-pm-info > div.am'
    ) as HTMLButtonElement;
    fireEvent.click(btn1);
  });
  test('should handle pm click', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={new Date('3/20/2022 4:00 AM')} onChange={testFn} />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const btn1 = container.querySelector(
      '.am-pm-info > div.pm'
    ) as HTMLButtonElement;
    fireEvent.click(btn1);
  });
  test('should handle hour change click', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={new Date('3/20/2022 4:00 AM')} onChange={testFn} />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const hour = container.querySelector('.position-2') as HTMLDivElement;
    fireEvent.click(hour);
  });
  test('should handle an id prop', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker
        time={new Date('3/20/2022 4:00 AM')}
        onChange={testFn}
        id={'1'}
      />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);
  });
  test('should handle minute change', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={new Date('3/20/2022 4:00 AM')} onChange={testFn} />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const hour = container.querySelector('.position-2') as HTMLDivElement;
    fireEvent.click(hour);

    const minute = container.querySelector('.position-9') as HTMLDivElement;
    fireEvent.click(minute);
  });
  test('should handle close', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker time={new Date('3/20/2022 4:00 AM')} onChange={testFn} />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const close = container.querySelector(
      '.c5-time-picker > div.clock > div.close > button'
    ) as HTMLButtonElement;
    fireEvent.click(close);
  });
  test('should handle return short time', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker
        time={new Date('3/20/2022 4:00 AM')}
        onChange={testFn}
        returnShortTime={true}
      />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const close = container.querySelector(
      '.c5-time-picker > div.clock > div.close > button'
    ) as HTMLButtonElement;
    fireEvent.click(close);
  });
  test('should handle return short time with an hour greater than 12', () => {
    const testFn = jest.fn();
    const { container, getByRole } = render(
      <TimePicker
        time={new Date('3/20/2022 4:00 PM')}
        onChange={testFn}
        returnShortTime={true}
      />
    );

    // open the picker
    const div = container.querySelector('.c5-tp-input') as HTMLDivElement;
    fireEvent.click(div);

    const close = container.querySelector(
      '.c5-time-picker > div.clock > div.close > button'
    ) as HTMLButtonElement;
    fireEvent.click(close);
  });
});
