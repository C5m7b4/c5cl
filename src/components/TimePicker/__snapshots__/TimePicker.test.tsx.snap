// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker should handle pm 1`] = `
<div>
  <div
    class="c5-time-picker"
    id="c5-time-picker"
  >
    <div
      class="c5-tp-input"
      id="c5-time-picker-1"
    >
      <input
        type="time"
      />
    </div>
  </div>
</div>
`;

exports[`TimePicker should render correctly 1`] = `
<div>
  <div
    class="c5-time-picker"
    id="c5-time-picker"
  >
    <div
      class="c5-tp-input"
      id="c5-time-picker-1"
    >
      <input
        type="time"
      />
    </div>
  </div>
</div>
`;
